rusk/src/lib/http/event.rs

```diff
@@ -0,0 +1,228 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use serde::{Deserialize, Serialize};
+use serde_with;
+use serde_with::serde_as;
+use std::fmt::{Display, Formatter};
+
+/// A request sent by the websocket client.
+#[derive(Debug, Serialize, Deserialize)]
+pub(crate) struct WsRequest {
+    pub(crate) headers: serde_json::Map<String, serde_json::Value>,
+    pub(crate) target: WsTarget,
+    // pub(crate) target_type: u8,
+    // pub(crate) target: String,
+    pub(crate) topic: String,
+    pub(crate) data: DataType,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) enum WsTarget {
+    Contract(String), // 0x01
+    Host(String),     // 0x02
+    Debugger(String), // 0x03
+}
+
+impl WsRequest {
+    pub fn x_headers(&self) -> serde_json::Map<String, serde_json::Value> {
+        let mut h = self.headers.clone();
+        h.retain(|k, _| k.to_lowercase().starts_with("x-"));
+        h
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Default)]
+pub struct WsResponse {
+    pub(crate) headers: serde_json::Map<String, serde_json::Value>,
+
+    /// The data returned by the contract call.
+    pub data: DataType,
+
+    /// A possible error happening during the contract call.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub error: Option<String>,
+}
+
+impl WsResponse {
+    pub fn from_error(error: String) -> Self {
+        Self {
+            error: Some(error),
+            ..Default::default()
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, Default)]
+#[serde(untagged)]
+pub enum DataType {
```

Comment:
This one I'm more unsure about:

```suggestion
pub enum Data {
```

Reply:
Me neither